### PR TITLE
Remove numpy from src/lightning/pytorch and use torch only

### PR DESCRIPTION
**Pull request by ishandutta0098**
## What does this PR do?
This PR is a part of #16649 
I have removed the numpy requirement from the files within `src/lightning/pytorch` including:
- `simple.py`
- `model_summary.py`

The numpy requirement has also been removed for `lr_finder.py`, it is in an older PR: #17264 

Both these PRs shall enable to replace numpy with torch for all files within `src/lightning/pytorch` 

<details>
  <summary><b>Before submitting</b></summary>

- [x] Was this **discussed/agreed** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/Lightning-AI/lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [x] Did you make sure to **update the documentation** with your changes? (if necessary)
- [ ] Did you write any **new necessary tests**? (not for typos and docs)
- [ ] Did you verify new and **existing tests pass** locally with your changes?
- [ ] Did you list all the **breaking changes** introduced by this pull request?
- [ ] Did you **update the CHANGELOG**? (not for typos, docs, test updates, or minor internal changes/refactors)

</details>

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing, make sure you have read the [review guidelines](https://github.com/Lightning-AI/lightning/wiki/Review-guidelines). In short, see the following bullet-list:

<details>
  <summary>Reviewer checklist</summary>

- [ ] Is this pull request ready for review? (if not, please submit in draft mode)
- [ ] Check that all items from **Before submitting** are resolved
- [ ] Make sure the title is self-explanatory and the description concisely explains the PR
- [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

</details>

### CODE REVIEW

File: src/lightning/pytorch/loggers/logger.py [hunk 1]
```diff
@@ -116,29 +116,53 @@ def merge_dicts(  # pragma: no cover
...
+    >>> agg_funcs = {'a': torch.mean, 'v': torch.max, 'd': {'d1': torch.sum}}
+    >>> result = merge_dicts([d1, d2, d3], agg_funcs, dflt_func)
+    >>> assert torch.allclose(result['a'], torch.tensor(1.3))
+    >>> assert torch.allclose(result['b'], torch.tensor(2.0))
+    >>> assert torch.allclose(result['c'], torch.tensor(1.0))
+    >>> assert torch.allclose(result['d']['d1'], torch.tensor(3.0))
+    >>> assert torch.allclose(result['d']['d2'], torch.tensor(3.0))
+    >>> assert torch.allclose(result['d']['d3'], torch.tensor(3.0))
+    >>> assert torch.allclose(result['d']['d4']['d5'], torch.tensor(1.0))
+    >>> assert torch.allclose(result['v'], torch.tensor(2.3))
```
Comment (Borda) [inline src/lightning/pytorch/loggers/logger.py:132]: doctest does not need assert, just pas the expected output

File: src/lightning/pytorch/utilities/model_summary/model_summary.py [hunk 2]
```diff
@@ -50,7 +49,7 @@ class LayerSummary:
         >>> model = torch.nn.Conv2d(3, 8, 3)
         >>> summary = LayerSummary(model)
         >>> summary.num_parameters
-        224
+        tensor(224.)
```
Comment (Borda) [inline src/lightning/pytorch/utilities/model_summary/model_summary.py:52]: expected output for doctests

Comment (ishandutta0098) [reply src/lightning/pytorch/utilities/model_summary/model_summary.py:52]: So I should ensure that the output is `224` and not `tensor(224.)` ?

File: src/lightning/pytorch/loggers/logger.py (hunk 1 above)
Comment (ishandutta0098) [reply src/lightning/pytorch/loggers/logger.py]: I have updated the doctest

File: src/lightning/pytorch/utilities/model_summary/model_summary.py (hunk 2 above)
Comment (ishandutta0098) [reply src/lightning/pytorch/utilities/model_summary/model_summary.py]: I have updated the doctest for this as well

File: src/lightning/pytorch/loggers/logger.py [hunk 3]
```diff
@@ -120,25 +120,50 @@ def merge_dicts(  # pragma: no cover
         >>> d1 = {'a': 1.7, 'b': 2.0, 'c': 1, 'd': {'d1': 1, 'd3': 3}}
         >>> d2 = {'a': 1.1, 'b': 2.2, 'v': 1, 'd': {'d1': 2, 'd2': 3}}
         >>> d3 = {'a': 1.1, 'v': 2.3, 'd': {'d3': 3, 'd4': {'d5': 1}}}
-        >>> dflt_func = min
-        >>> agg_funcs = {'a': np.mean, 'v': max, 'd': {'d1': sum}}
+        >>> dflt_func = torch.min
+        >>> agg_funcs = {'a': torch.mean, 'v': torch.max, 'd': {'d1': torch.sum}}
```
Comment (aniketmaurya) [inline src/lightning/pytorch/loggers/logger.py:124]: `max` and `sum` seemed to be Python functions and not numpy

File: src/lightning/pytorch/loggers/logger.py [hunk 4]
```diff
@@ -94,7 +94,7 @@ def method(*args: Any, **kwargs: Any) -> None:
 def merge_dicts(  # pragma: no cover
     dicts: Sequence[Mapping],
     agg_key_funcs: Optional[Mapping] = None,
-    default_func: Callable[[Sequence[float]], float] = np.mean,
```
Comment (aniketmaurya) [inline src/lightning/pytorch/loggers/logger.py:97]: `Callable[[Sequence[float]], float]` seems to be Python built-in `float` and not numpy

File: src/lightning/pytorch/loggers/logger.py [hunk 5]
```diff
@@ -120,25 +120,50 @@ def merge_dicts(  # pragma: no cover
...
-        >>> agg_funcs = {'a': np.mean, 'v': max, 'd': {'d1': sum}}
+        >>> dflt_func = torch.min
+        >>> agg_funcs = {'a': torch.mean, 'v': torch.max, 'd': {'d1': torch.sum}}
         >>> pprint.pprint(merge_dicts([d1, d2, d3], agg_funcs, dflt_func))
-        {'a': 1.3,
-         'b': 2.0,
-         'c': 1,
-         'd': {'d1': 3, 'd2': 3, 'd3': 3, 'd4': {'d5': 1}},
-         'v': 2.3}
+        {'a': tensor(1.3000),
```
Comment (aniketmaurya) [inline src/lightning/pytorch/loggers/logger.py:126]:
```
{'a': 1.3,
         'b': 2.0,
         'c': 1,
         'd': {'d1': 3, 'd2': 3, 'd3': 3, 'd4': {'d5': 1}},
         'v': 2.3}
```

these variables were Python builtin, not sure if we want to replace that with torch.Tensor. @Borda ?

File: src/lightning/pytorch/strategies/launchers/multiprocessing.py [hunk 6]
```diff
@@ -210,7 +209,6 @@ def _check_torchdistx_support(self) -> None:
 
     def get_extra_results(self, trainer: "pl.Trainer") -> Dict[str, Any]:
         """Gather extra state from the Trainer and return it as a dictionary for sending back to the main process.
-        To avoid issues with memory sharing, we cast the data to numpy.
```
Comment (aniketmaurya) [inline src/lightning/pytorch/strategies/launchers/multiprocessing.py:213]: casting to numpy is done to prevent issues with memory sharing so I won't change this as a part of the refactor PR.

File: src/lightning/pytorch/utilities/model_summary/model_summary.py [hunk 7]
```diff
@@ -120,8 +119,13 @@ def layer_type(self) -> str:
     @property
     def num_parameters(self) -> int:
         """Returns the number of parameters in this module."""
-        return sum(
-            cast(int, np.prod(p.shape)) if not _is_lazy_weight_tensor(p) else 0 for p in self._module.parameters()
+        return int(
```
Comment (aniketmaurya) [inline src/lightning/pytorch/utilities/model_summary/model_summary.py:122]: why do we need to cast `int` again if sum values are already `int`

Comment (ishandutta0098) [reply src/lightning/pytorch/utilities/model_summary/model_summary.py:122]: It returns a tensor as Borda mentioned in the above conversation, but in the doctest it returns an int. Hence I made it int manually.

File: src/lightning/pytorch/strategies/launchers/multiprocessing.py (hunk 6 above)
Comment (ishandutta0098) [reply src/lightning/pytorch/strategies/launchers/multiprocessing.py:213]: Okay I will update this

Comment (ishandutta0098) [reply src/lightning/pytorch/strategies/launchers/multiprocessing.py]: I have reverted the changes for this file

File: src/lightning/pytorch/loggers/logger.py (hunk 3 above)
Comment (ishandutta0098) [reply src/lightning/pytorch/loggers/logger.py]: Committed the changes

File: src/lightning/pytorch/loggers/logger.py (hunk 4 above)
Comment (ishandutta0098) [reply src/lightning/pytorch/loggers/logger.py:97]: Can you clarify this further @aniketmaurya ?

File: src/lightning/pytorch/utilities/model_summary/model_summary.py [hunk 8]
```diff
@@ -392,8 +396,8 @@ def get_human_readable_count(number: int) -> str:
     """
     assert number >= 0
     labels = PARAMETER_NUM_UNITS
-    num_digits = int(np.floor(np.log10(number)) + 1 if number > 0 else 1)
-    num_groups = int(np.ceil(num_digits / 3))
+    num_digits = int(torch.floor(torch.log10(torch.tensor(number, dtype=torch.float32))) + 1 if number > 0 else 1)
+    num_groups = int(torch.ceil(torch.tensor(num_digits, dtype=torch.float32) / 3))
```
Comment (awaelchli) [inline src/lightning/pytorch/utilities/model_summary/model_summary.py:400]: We don't need to convert to tensors here. Let's simplify it like this?
```suggestion
    num_digits = int(math.floor(math.log10(number)) + 1 if number > 0 else 1)
    num_groups = int(math.ceil(num_digits / 3))
```

File: src/lightning/pytorch/utilities/model_summary/model_summary.py [hunk 9]
```diff
@@ -120,8 +119,13 @@ def layer_type(self) -> str:
     @property
     def num_parameters(self) -> int:
         """Returns the number of parameters in this module."""
-        return sum(
-            cast(int, np.prod(p.shape)) if not _is_lazy_weight_tensor(p) else 0 for p in self._module.parameters()
+        return int(
+            sum(
+                cast(int, torch.prod(torch.tensor(p.shape, dtype=torch.float32)))
```
Comment (awaelchli) [inline src/lightning/pytorch/utilities/model_summary/model_summary.py:124]:
```suggestion
                cast(int, math.prod(p.shape))
```

File: src/lightning/pytorch/profilers/simple.py [hunk 10]
```diff
@@ -80,15 +80,24 @@ def stop(self, action_name: str) -> None:
...
         report = [
-            (a, np.mean(d), len(d), np.sum(d), 100.0 * np.sum(d) / total_duration)
+            (
+                a,
+                torch.mean(torch.tensor(d)).item(),
+                len(d),
+                torch.sum(torch.tensor(d)).item(),
+                100.0 * torch.sum(torch.tensor(d)).item() / total_duration,
+            )
             for a, d in self.recorded_durations.items()
```
Comment (awaelchli) [inline src/lightning/pytorch/profilers/simple.py:90]: @ishandutta0098 what is the type of the variables a and d here?

File: src/lightning/pytorch/loggers/logger.py [hunk 11]
```diff
@@ -121,24 +121,46 @@ def merge_dicts(  # pragma: no cover
...
-        >>> agg_funcs = {'a': np.mean, 'v': max, 'd': {'d1': sum}}
+        >>> agg_funcs = {'a': torch.mean, 'v': torch.max, 'd': {'d1': torch.sum}}
         >>> pprint.pprint(merge_dicts([d1, d2, d3], agg_funcs, dflt_func))
         {'a': 1.3,
          'b': 2.0,
          'c': 1,
          'd': {'d1': 3, 'd2': 3, 'd3': 3, 'd4': {'d5': 1}},
          'v': 2.3}
     """
+    # If agg_key_funcs is not provided, initialize it as an empty dictionary
```
[the remark this answers is not in the document]
Comment (ishandutta0098) [reply src/lightning/pytorch/loggers/logger.py]: I have removed them now

File: src/lightning/pytorch/profilers/simple.py (hunk 10 above)
Comment (ishandutta0098) [reply src/lightning/pytorch/profilers/simple.py:90]: `a` is `str` and `d` is `list`

File: src/lightning/pytorch/utilities/model_summary/model_summary.py (hunk 7 above)
Comment (ishandutta0098) [reply src/lightning/pytorch/utilities/model_summary/model_summary.py]: I have updated the code and removed the cast to `int`

File: src/lightning/pytorch/profilers/simple.py [hunk 12]
```diff
@@ -80,15 +80,24 @@ def stop(self, action_name: str) -> None:
     def _make_report_extended(self) -> Tuple[_TABLE_DATA_EXTENDED, float, float]:
         total_duration = time.monotonic() - self.start_time
         report = [
-            (a, np.mean(d), len(d), np.sum(d), 100.0 * np.sum(d) / total_duration)
+            (
+                a,
+                torch.mean(torch.tensor(d)).item(),
```
Comment (carmocca) [inline src/lightning/pytorch/profilers/simple.py:85]: Can you convert d to tensor just once? You'll need to convert this list comprehension into a loop

File: src/lightning/pytorch/profilers/simple.py [hunk 13]
```diff
@@ -80,15 +80,24 @@ def stop(self, action_name: str) -> None:
...
             for a, d in self.recorded_durations.items()
         ]
         report.sort(key=lambda x: x[4], reverse=True)
         total_calls = sum(x[2] for x in report)
         return report, total_calls, total_duration
 
     def _make_report(self) -> _TABLE_DATA:
-        report = [(action, np.mean(d), np.sum(d)) for action, d in self.recorded_durations.items()]
+        report = [
+            (action, torch.mean(torch.tensor(d)).item(), torch.sum(torch.tensor(d)).item())
```
Comment (carmocca) [inline src/lightning/pytorch/profilers/simple.py:98]: Same suggestion

File: src/lightning/pytorch/loggers/logger.py [hunk 14]
```diff
@@ -111,10 +109,8 @@ def merge_dicts(  # pragma: no cover
         default_func:
             Default function to aggregate keys, which are not presented in the
             `agg_key_funcs` map.
-
     Returns:
         Dictionary with merged values.
-
```
Comment (carmocca) [inline src/lightning/pytorch/loggers/logger.py:117]: The changes to this file should be reverted. They don't seem to relate to the PR's purpose

File: src/lightning/pytorch/profilers/simple.py [hunk 15]
```diff
@@ -79,16 +79,30 @@ def stop(self, action_name: str) -> None:
...
         total_duration = time.monotonic() - self.start_time
-        report = [
-            (a, np.mean(d), len(d), np.sum(d), 100.0 * np.sum(d) / total_duration)
-            for a, d in self.recorded_durations.items()
-        ]
+        report = []
+
+        for a, d in self.recorded_durations.items():
+            d_tensor = torch.tensor(d)
+            mean_d = torch.mean(d_tensor).item()
```
Comment (awaelchli) [inline src/lightning/pytorch/profilers/simple.py:86]: It is not necessary to compute the mean separately if we anyway havae to compute the sum below.

File: src/lightning/pytorch/profilers/simple.py [hunk 16]
```diff
@@ -79,16 +79,30 @@ def stop(self, action_name: str) -> None:
...
         total_calls = sum(x[2] for x in report)
         return report, total_calls, total_duration
 
     def _make_report(self) -> _TABLE_DATA:
-        report = [(action, np.mean(d), np.sum(d)) for action, d in self.recorded_durations.items()]
+        report = []
+        for action, d in self.recorded_durations.items():
+            d_tensor = torch.tensor(d)
+            mean_d = torch.mean(d_tensor).item()
+            sum_d = torch.sum(d_tensor).item()
```
Comment (awaelchli) [inline src/lightning/pytorch/profilers/simple.py:102]: same here :)

**Review by awaelchli — APPROVED** [review]

File: src/lightning/pytorch/profilers/simple.py (hunk 15 above)
Comment (ishandutta0098) [reply src/lightning/pytorch/profilers/simple.py]: Yes you are right, I have updated and committed

File: src/lightning/pytorch/profilers/simple.py (hunk 16 above)
Comment (ishandutta0098) [reply src/lightning/pytorch/profilers/simple.py]: I have updated and committed

File: src/lightning/pytorch/utilities/model_summary/model_summary.py [hunk 17]
```diff
@@ -121,7 +122,7 @@ def layer_type(self) -> str:
     def num_parameters(self) -> int:
         """Returns the number of parameters in this module."""
         return sum(
-            cast(int, np.prod(p.shape)) if not _is_lazy_weight_tensor(p) else 0 for p in self._module.parameters()
+            cast(int, math.prod(p.shape)) if not _is_lazy_weight_tensor(p) else 0 for p in self._module.parameters()
```
Comment (carmocca) [inline src/lightning/pytorch/utilities/model_summary/model_summary.py:125]:
```suggestion
            math.prod(p.shape) if not _is_lazy_weight_tensor(p) else 0 for p in self._module.parameters()
```

**Review by Borda — APPROVED** [review]
